### ISSUE TITLE
Optimize memory profile

### ISSUE THREAD

**Issue by dbkaplun** [balderdashy/waterline#805]
I am not too familiar with the waterline source code but it's likely that waterline is buffering the entire database response before formatting it and sending it off. A thousand rows occupies about 50MB RAM which could be optimized by streaming.

**Comment by andyhu** [comment]
Yes, that's also my concern.  But this might be an issue for individual database adapters rather than waterline itself. I'm a bit surprised that when doing an update with `sails-mongo` adapter, it will query the database 3 times, 2 `find` operations and one `update` operation, in order to return the full list of the updated result. But this is really not necessary in most cases and if there're let's say 10, 000 documents with lots of fields need to be updated, the returned result will be huge in memory. Can we add a option in connections.js, if turned on, when doing update, only return number of documents updated rather than the whole list of all documents has been updated? If so, I'm more than happy to contribute PRs for making it happen sooner.

**Comment by andyhu** [comment]
See the code here for how it's implemented:
https://github.com/balderdashy/sails-mongo/blob/master/lib/collection.js#L196

**Comment by particlebanana** [comment]
I'm all about streaming results! The current version of WL has a ton of optimizations needed. In the case of the delete it's to match the api that was written before we had a ton of adapters. So the api and tests expect to receive the results of the delete, which is trivial in something like postgresql but requires an extra query in mongo. Basically if you want to have the ability to change adapters and not change code you end up with inefficiencies in all or most adapters.

We won't get there in this WL codebase but if you guys are really interested in helping with this stuff please take a look at WL2 (https://github.com/mikermcneil/waterline2) as it's our playground for a complete new api. If we can get streaming in there that would be fantastic.

**Comment by andyhu** [comment]
Thanks @particlebanana for the clarification. Will certainly check WL2!

**Comment by particlebanana** [comment]
Also just an update most adapters support streaming queries with the `.stream()` method. For now it only supports finds and not updates/deletes.

**Comment by tjwebb** [comment]
One way to get some quick and cheap memory savings is to set `dynamicFinders: false` in your models config.

https://github.com/balderdashy/waterline/pull/594#issuecomment-52437235

**Comment by sailsbot** [comment]
Thanks for posting, @dbkaplun.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by mikermcneil** [comment]
@dbkaplun there is built-in experimental support for `.stream()` currently-- we implemented it originally to support streaming large XML sitemaps (i.e. in order to fulfill the maximum page size for google).  But realistically, for the vast majority of use cases, you don't want to be fetching thousands of rows at a time-- you want to be fetching only as much data as you need to populate the UI.  Hope that helps!
